client/reader/controller.js

```diff
@@ -55,6 +57,23 @@ function trackScrollPage( path, title, category, readerView, pageNum ) {
 	} );
 }
 
+function onRouteChange( newContext, oldContext ) {
+	if ( ! oldContext ) {
+		return;
+	}
+
+	const fullPostViewPrefix = '/read/post/';
+
+	if ( startsWith( oldContext.path, fullPostViewPrefix ) &&
+		! startsWith( newContext.path, fullPostViewPrefix ) &&
+		__fullPostInstance ) {
+		__fullPostInstance.setState( { isVisible: false } );
+		__fullPostInstance = false;
+	}
+}
+
+routeChange( onRouteChange );
```

Comment:
Unattached named functions provide a better stack trace, while anonymous fat arrow is more succinct and gives us lexical this. I don't really have a :horse: in this, as both look fine to me in this case.